cmake/modules/extensions.cmake

```diff
@@ -3328,7 +3328,7 @@ function(zephyr_check_cache variable)
     # without cleaning first
     if(cli_argument)
       if(NOT ((CACHED_${variable} STREQUAL cli_argument) OR (${variable}_DEPRECATED STREQUAL cli_argument)))
-        message(WARNING "The build directory must be cleaned pristinely when "
+        message(FATAL_ERROR "The build directory must be cleaned pristinely when "
```

Comment:
this is not only impacting snippets but is impacting every setting which cannot be changed, including `BOARD` and SHIELD`. 

Thus I believe a broader audience should be included if we're to change the build system behavior.

Also. if this is changed to an error, then the following code, which becomes dead code with the proposed change, should be cleaned up in addition.

Reply:
I'll have a look at cleaning up the following code. Also changed the title to take into account your feedback.

Reply:
another, and perhaps better alternative, is to change from `WARNING` to `AUTHOR_WARNING`.

That will allow developers to error on warnings by calling cmake with: `-Werror=dev`, while still keeping it a warning for others.

Reply:
By the way, an error is already thrown by West when compiling for a different target board without a pristine build:

![image](https://github.com/user-attachments/assets/b2a36a44-8d87-4449-b0fe-6ca15568702e)

But it will compile with a warning with cmake directly. Shouldn't we be consistent between cmake and west?

Reply:
not sure why `west build` decided to error out early on board, but `zephyr_check_cache` is a common available Zephyr CMake extension function which can used elsewhere, even by downstream projects / modules.